### PR TITLE
Add supplier to EnumClassInfo

### PR DESCRIPTION
**Pull request by Fusezion**
### Description
This PR aims to update the EnumClassInfo with a provided supplier method, this way it's no dependant on the creation of one within the `ClassInfo#getSupplier` method.

In addition to this change EnumClassInfo parameters were changed to more understandable ones, rather than stopping there I added StringMode support into `EnumUtils#toString` this change was done mostly for ease of use of developers that may or may not change based off flags, since flags already represented the StringMode I wanted to increase usability in these. Since existing implementations lack the support for it.

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** none

### CODE REVIEW

**Comment by sovdeeth** [comment]
I'm not against this change but what exactly is the benefit here? Shouldn't users never interact with supplier directly, and only through getSupplier?

**Comment by Fusezion** [comment]
> I'm not against this change but what exactly is the benefit here? Shouldn't users never interact with supplier directly, and only through getSupplier?

There is likely no benefit outside of at least understanding where things are taking place, a user shouldn't need to feel a need to add `.supplier()` to an EnumClassInfo like it was recently in #7086 before being removed once mentioned.

getSupplier should only return a supplier not set a supplier if it's not already set

**Comment by sovdeeth** [comment]
> > I'm not against this change but what exactly is the benefit here? Shouldn't users never interact with supplier directly, and only through getSupplier?
> 
> There is likely no benefit outside of at least understanding where things are taking place, a user shouldn't need to feel a need to add `.supplier()` to an EnumClassInfo like it was recently in #7086 before being removed once mentioned.
> 
> getSupplier should only return a supplier not set a supplier if it's not already set

well it's rather beneficial to not have a supplier until it's actually needed, but that's pretty minor
I think you should add the explicit supplier to enumclassinfos but keep the getSupplier method as is
No need to potentially break things for no benefit

**Review by sovdeeth — APPROVED** [review]

File: src/main/java/ch/njol/skript/classes/EnumClassInfo.java [hunk 1]
```diff
@@ -33,41 +16,42 @@
...
-			.serializer(new EnumSerializer<>(c))
+			.serializer(new EnumSerializer<>(enumClass))
 			.defaultExpression(defaultExpression)
+			.supplier(() -> new ArrayIterator<>(enumClass.getEnumConstants()))
 			.parser(new Parser<T>() {
 				@Override
 				@Nullable
-				public T parse(String s, ParseContext context) {
-					return enumUtils.parse(s);
+				public T parse(String input, ParseContext context) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/classes/EnumClassInfo.java:43]:
```suggestion
				public @Nullable T parse(String input, ParseContext context) {
```

**Review by Efnilite — APPROVED** [review]

**Review by APickledWalrus — APPROVED** [review]

**Pull request merged**